_examples/sql/transactional-events/main.go

```diff
@@ -0,0 +1,171 @@
+package main
+
+import (
+	"context"
+	stdSQL "database/sql"
+	"encoding/json"
+	"log"
+	"time"
+
+	driver "github.com/go-sql-driver/mysql"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill-kafka/pkg/kafka"
+	"github.com/ThreeDotsLabs/watermill-sql/pkg/sql"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/ThreeDotsLabs/watermill/message/router/middleware"
+	"github.com/ThreeDotsLabs/watermill/message/router/plugin"
+)
+
+var (
+	logger     = watermill.NewStdLogger(false, false)
+	kafkaTopic = "events"
+	mysqlTable = "events"
+)
+
+func main() {
+	router, err := message.NewRouter(message.RouterConfig{}, logger)
+	if err != nil {
+		panic(err)
+	}
+
+	router.AddPlugin(plugin.SignalsHandler)
+	router.AddMiddleware(middleware.Recoverer)
+
+	db := createDB()
+
+	subscriber := createSubscriber(db)
+	publisher := createPublisher()
+
+	router.AddHandler(
+		"mysql-to-kafka",
+		mysqlTable,
+		subscriber,
+		kafkaTopic,
+		publisher,
+		func(msg *message.Message) ([]*message.Message, error) {
+			consumedEvent := event{}
+			err := json.Unmarshal(msg.Payload, &consumedEvent)
+			if err != nil {
+				return nil, err
+			}
+
+			log.Printf("received event %+v with UUID %s", consumedEvent, msg.UUID)
+
+			return []*message.Message{msg}, nil
+		},
+	)
+
+	go func() {
+		<-router.Running()
+		simulateEvents(db)
+	}()
+
+	if err := router.Run(context.Background()); err != nil {
+		panic(err)
+	}
+}
+
+func createDB() *stdSQL.DB {
+	conf := driver.NewConfig()
+	conf.Net = "tcp"
+	conf.User = "root"
+	conf.Addr = "mysql"
+	conf.DBName = "watermill"
+
+	db, err := stdSQL.Open("mysql", conf.FormatDSN())
+	if err != nil {
+		panic(err)
+	}
+
+	err = db.Ping()
+	if err != nil {
+		panic(err)
+	}
+
+	return db
+}
+
+func createSubscriber(db *stdSQL.DB) message.Subscriber {
+	pub, err := sql.NewSubscriber(
+		db,
+		sql.SubscriberConfig{
+			SchemaAdapter:    sql.DefaultSchema{},
+			OffsetsAdapter:   sql.DefaultMySQLOffsetsAdapter{},
+			InitializeSchema: true,
+		},
+		logger,
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	return pub
+}
+
+func createPublisher() message.Publisher {
+	pub, err := kafka.NewPublisher(
+		[]string{"kafka:9092"},
+		kafka.DefaultMarshaler{},
+		nil,
+		logger,
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	return pub
+}
+
+type event struct {
+	Name       string `json:"name"`
+	OccurredAt string `json:"occurred_at"`
+}
+
+func simulateEvents(db *stdSQL.DB) {
+	for {
+		tx, err := db.Begin()
+		if err != nil {
+			panic(err)
+		}
+
+		err = publishEvent(tx)
+		if err != nil {
+			rollbackErr := tx.Rollback()
+			if rollbackErr != nil {
+				panic(rollbackErr)
+			}
+			panic(err)
+		}
+
+		err = tx.Commit()
+		if err != nil {
+			panic(err)
+		}
+
+		time.Sleep(time.Second)
+	}
+}
+
+func publishEvent(tx *stdSQL.Tx) error {
+	pub, err := sql.NewPublisher(tx, sql.PublisherConfig{
```

Comment:
we could use a comment like 
`// we create a new publisher for each event, in order for the event publication to be transactional`, or something better worded?

Reply:
Added a comment in 131b47695b6334095112ccef6d08d3f9e8ab741f

Reply:
perfect <3